### PR TITLE
FEAT / Bitsandbytes: Add `dequantize` API for bitsandbytes quantized models

### LINKED ISSUE #30177: Load nf4 weights/model in bfloat16
### Feature request

Allow quantized models to be loaded (upcasted) to bfloat16.

### Motivation

1. The big benefit of QLoRA is being able to fine-tune on less VRAM AND with much less time than full fine-tuning.
2. Adding in rsLoRA, initiating LoRA adapters to minimise quanting error,  and training norm and embedding layers gets performance close to full fine-tuning.

HOWEVER, there's a catch that it's not possible to get back to a bf16 model (yes, it's possible to merge to 4-bit, but this hurts accuracy, and there's no way to upcast either). See [this issue](https://github.com/huggingface/peft/pull/851).

### Your contribution

BEST approach:
- Allow a type of merge where a quantized base model is merged with LoRA weights to form a 16 bit (float16 or bfloat16) model. [Obviously this will take more RAM or VRAM, so may not suit certain users, but it is best for quality].

HACKY but sufficient approach:
- Simply allow a quantized model (i.e. saved like [this](https://github.com/huggingface/transformers/pull/26037)] to be loaded in bfloat 16. If that's possible, then the following flow would work:
1. Fine tune a model with QLoRA (ideally making norms and embeds trainable). [initialise the adapters to minimise quanting error]
2. Save adapters.
3. Unload the adapters and save the nf4 base model (ideally including the trained norms and embeds)
4. Re-load the nf4 base model, but upcasted to bfloat16.
5. Apply the adapter
6. Merge the adapter to the base model
7. This should result in a highly accurate final model.

Everything 1-7 is possible already except for step 4. However, step 4 is done in unsloth, see [here](https://github.com/unslothai/unsloth/issues/195), so should be fairly easy.

[Side note - I'm unsure whether bnb quants the norm and embed layers, or just the linear layers?]

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/30177

This PR adds a new feature `dequantize` in order to de-quantize models for interesting usecases such as the one described in #30177 

The API is very simple:

```python
from transformers import AutoModelForCausalLM, BitsAndBytesConfig, AutoTokenizer

model_id = "facebook/opt-125m"

model = AutoModelForCausalLM.from_pretrained(model_id, quantization_config=BitsAndBytesConfig(load_in_4bit=True))
tokenizer = AutoTokenizer.from_pretrained(model_id)

model.dequantize()

text = tokenizer("Hello my name is", return_tensors="pt").to(0)

out = model.generate(**text)
print(tokenizer.decode(out[0]))
```

Users just need to make sure they have enough GPU RAM in order to store the unquantized model, otherwise they might face unexpected behaviour

Added the support for 4-bit / 8-bit models and nice tests + docs to educate users on how to use this new API.

cc @amyeroberts @SunMarc

### CODE REVIEW

File: src/transformers/integrations/bitsandbytes.py [hunk 1]
```diff
@@ -322,3 +325,122 @@ def get_keys_to_not_convert(model):
...
+        raise TypeError(f"Input weight should be of type nn.Parameter, got {type(weight)} instead")
+
+    cls_name = weight.__class__.__name__
+    if cls_name not in ("Params4bit", "Int8Params"):
+        return weight
+
+    import bitsandbytes as bnb
+
+    if cls_name == "Params4bit":
+        return bnb.functional.dequantize_4bit(weight.data, weight.quant_state)
```
Comment (SunMarc) [inline src/transformers/integrations/bitsandbytes.py:347]: The user might want to know in which precision the model was dequantized since they don't have the possibility to control that. I think it could be great to give that information since there is no default value (as opposed to `from_pretrained` which loads the model in fp32). 
Two ways to get that: 
- just check the dtype of the weights at the end ( potentially the easiest way ) 
- check that happens in `dequantize_4bit` . In the method, you see that they get the output dtype with weight.quant_state.dtype. 

We can potentially add a `torch_dtype` attribute in the future if it makes sense.

**Review by SunMarc — APPROVED** [review]
Thanks for adding this new method in quantizer ! This will make fine-tuning with quantized model way easier ! I left a few minor comments.

File: tests/quantization/bnb/test_mixed_int8.py [hunk 2]
```diff
@@ -285,6 +285,23 @@ def test_generate_quality_config(self):
 
         self.assertIn(self.tokenizer.decode(output_sequences[0], skip_special_tokens=True), self.EXPECTED_OUTPUTS)
 
+    def test_generate_quality_dequantize(self):
+        r"""
+        Test that loading the model and unquantize it produce correct results
```
Comment (amyeroberts) [inline tests/quantization/bnb/test_mixed_int8.py:290]:
```suggestion
        Test that loading the model and dequantizing it produce correct results
```

File: src/transformers/integrations/bitsandbytes.py (hunk 1 above)
Comment (amyeroberts) [reply src/transformers/integrations/bitsandbytes.py:347]: +1

File: src/transformers/integrations/bitsandbytes.py [hunk 3]
```diff
@@ -322,3 +325,122 @@ def get_keys_to_not_convert(model):
...
+    has_been_replaced=False,
+):
+    """
+    Converts a quantized model into its unquantized original version. The newly converted model will have
+    some performance drop compared to the original model before quantization - use it only for specific usecases
+    such as QLoRA adapters merging.
+
+    Returns the converted model and a boolean that indicates if the conversion has been successfull or not.
+    """
+    import bitsandbytes as bnb
```
Comment (amyeroberts) [inline src/transformers/integrations/bitsandbytes.py:392]: This is already imported at the top of the module

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this! +1 on all of @SunMarc's comments.

File: src/transformers/integrations/bitsandbytes.py (hunk 1 above)
Comment (younesbelkada) [reply src/transformers/integrations/bitsandbytes.py]: Nice catch! The output dtype should be correctly inferred here: https://github.com/TimDettmers/bitsandbytes/blob/b891f80ba514833f41f0e9226983b02a9fb5c44b/bitsandbytes/functional.py#L1349 through the compute_dtype so it should be accurate - I added a `warning_once` staement to inform users on the dequantized dtype: [1a4a906](https://github.com/huggingface/transformers/pull/30806/commits/1a4a906741c2be1bab39538d9ad117e12d3d67f9)

File: src/transformers/integrations/bitsandbytes.py (hunk 3 above)
Comment (younesbelkada) [reply src/transformers/integrations/bitsandbytes.py]: Nice catch ! Should be fixed now

File: src/transformers/modeling_utils.py [hunk 4]
```diff
@@ -1327,6 +1327,14 @@ def post_init(self):
         self.init_weights()
         self._backward_compatibility_gradient_checkpointing()
 
+    def dequantize(self):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/modeling_utils.py:1330]: Done!

File: src/transformers/quantizers/base.py [hunk 5]
```diff
@@ -194,6 +194,23 @@ def postprocess_model(self, model: "PreTrainedModel", **kwargs):
         """
         return self._process_model_after_weight_loading(model, **kwargs)
 
+    def dequantize(self, model):
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/quantizers/base.py:197]: Done ç

File: src/transformers/quantizers/base.py [hunk 6]
```diff
@@ -194,6 +194,23 @@ def postprocess_model(self, model: "PreTrainedModel", **kwargs):
...
+        if self.requires_calibration:
+            raise ValueError(
+                f"{self.quantization_config.quant_method} requires data calibration, it cannot be dequantized."
+            )
```
[the remark this answers is not in the document]
Comment (younesbelkada) [reply src/transformers/quantizers/base.py:205]: Thanks for the heads up ! Yes as we discussed offline, this makes sense!

File: src/transformers/integrations/bitsandbytes.py [hunk 7]
```diff
@@ -322,3 +325,122 @@ def get_keys_to_not_convert(model):
...
+            _, has_been_replaced = dequantize_and_replace(
+                module,
+                modules_to_not_convert,
+                current_key_name,
+                quantization_config,
+                has_been_replaced=has_been_replaced,
+            )
+        # Remove the last key for recursion
+        current_key_name.pop(-1)
+    return model, has_been_replaced
```
Comment (amyeroberts) [inline src/transformers/integrations/bitsandbytes.py:446]: One general comment, if instead you have a private method `_dequantize_and_replace`, which handles the recursion, you don't need to return `has_been_replaced` here. When someone calls `dequantize_and_replace`, I don't think `has_been_replaced` is ever used and could be confusing e.g.: 

```
# This is just dequantize_and_replace from before
def _dequantize_and_replace(
    model,
    modules_to_not_convert=None,
    current_key_name=None,
    quantization_config=None,
    has_been_replaced=False,
):
    ...
    return model, has_been_replaced

def dequantize_and_replace(
    model,
    modules_to_not_convert=None,
    current_key_name=None,
    quantization_config=None,
    has_been_replaced=False,
):
    model, has_been_replaced = _dequantize_and_replace(...)
    return model 
```

File: src/transformers/integrations/bitsandbytes.py [hunk 8]
```diff
@@ -322,3 +325,122 @@ def get_keys_to_not_convert(model):
...
+
+def dequantize_and_replace(
+    model,
+    modules_to_not_convert=None,
+    current_key_name=None,
+    quantization_config=None,
+    has_been_replaced=False,
+):
+    """
+    Converts a quantized model into its unquantized original version. The newly converted model will have
```
Comment (amyeroberts) [inline src/transformers/integrations/bitsandbytes.py:388]:
```suggestion
    Converts a quantized model into its dequantized original version. The newly converted model will have
```

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this feature and iterating!

File: src/transformers/integrations/bitsandbytes.py (hunk 7 above)
Comment (younesbelkada) [reply src/transformers/integrations/bitsandbytes.py:446]: Done in https://github.com/huggingface/transformers/pull/30806/commits/8b904f7641850d24005e4a00ba72e17e03c62519 !

File: src/transformers/integrations/bitsandbytes.py [hunk 9]
```diff
@@ -444,3 +444,22 @@ def dequantize_and_replace(
...
+    quantization_config=None,
+):
+    model, has_been_replaced = _dequantize_and_replace(
+        model,
+        modules_to_not_convert=modules_to_not_convert,
+        quantization_config=quantization_config,
+    )
+
+    if not has_been_replaced:
+        logger.warning(
```
Comment (amyeroberts) [inline src/transformers/integrations/bitsandbytes.py:461]: Nice :)

**Comment by RonanKMcGovern** [comment]
Yeah this is great, thanks

**Comment by younesbelkada** [comment]
Great thanks @RonanKMcGovern ! let us know how it goes
